internal/scheduler/legacyscheduler.go

```diff
@@ -317,99 +358,248 @@ func (it *QueueCandidateGangIterator[T]) schedulingReportsFromJobs(ctx context.C
 	return reports, nil
 }
 
+func totalResourceRequestsFromJobs(jobs []LegacySchedulerJob, priorityClasses map[string]configuration.PriorityClass) schedulerobjects.ResourceList {
+	rv := schedulerobjects.ResourceList{}
+	for _, job := range jobs {
+		for _, reqs := range job.GetRequirements(priorityClasses).GetObjectRequirements() {
+			rv.Add(
+				schedulerobjects.ResourceListFromV1ResourceList(
+					reqs.GetPodRequirements().ResourceRequirements.Requests,
+				),
+			)
+		}
+	}
+	return rv
+}
+
+// Priority queue used by CandidateGangIterator to determine from which queue to schedule the next job.
+type QueueCandidateGangIteratorPQ []*QueueCandidateGangIteratorItem
+
+type QueueCandidateGangIteratorItem struct {
+	// Each item corresponds to a queue.
+	queue string
+	// Iterator for this queue.
+	it *QueueCandidateGangIterator
+	// Most recent value produced by the iterator.
+	// Cached here to avoid repeating scheduling checks unnecessarily.
+	v []*JobSchedulingReport
+	// Fraction of its fair share this queue would have
+	// if its next schedulable job were to be scheduled.
+	fractionOfFairShare float64
+	// The index of the item in the heap.
+	// maintained by the heap.Interface methods.
+	index int
+}
+
+func (pq QueueCandidateGangIteratorPQ) Len() int { return len(pq) }
+
+func (pq QueueCandidateGangIteratorPQ) Less(i, j int) bool {
+	// Tie-break by queue name.
+	if pq[i].fractionOfFairShare == pq[j].fractionOfFairShare {
+		return pq[i].queue < pq[j].queue
+	}
+	return pq[i].fractionOfFairShare < pq[j].fractionOfFairShare
+}
+
+func (pq QueueCandidateGangIteratorPQ) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *QueueCandidateGangIteratorPQ) Push(x any) {
+	n := len(*pq)
+	item := x.(*QueueCandidateGangIteratorItem)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+func (pq *QueueCandidateGangIteratorPQ) Pop() any {
```

Comment:
will this panic if you try  and Pop from an empty queue?

Reply:
Just to be clear, I'd expect that popping from an empty queue would be a valid operation and would return nil (assuming reference types only) or a (zeroValue, bool) (assuming we can have value types too)

Reply:
Yes, it'll panic on calling `Pop()` on an empty queue. I think doing so is analogous to trying to access an out of bounds index on a slice and that it should cause a panic.

Reply:
Additionally, this is the standard PQ implementation given in the docs. Not that we should necessarily use that implementation if we think it's wrong though.
https://pkg.go.dev/container/heap#example-package-PriorityQueue